src/main/java/org/squiddev/plethora/integration/astralsorcery/IntegrationAstralSorcery.java

```diff
@@ -0,0 +1,777 @@
+		ItemSkyResonator.class,
+		"Provides the available modes for this Resonator"
+	) {
+		@Nonnull
+		@Override
+		public Map<String, Object> getMeta(@Nonnull IPartialContext<ItemStack> context, @Nonnull ItemSkyResonator item) {
+			//REFINE Do we want `getUnlocalizedName` or `getUnlocalizedUpgradeName`?
+			// Further, should we localize the String?
+
+			/*
+			ItemStack stack = context.getTarget();
+			List<ItemSkyResonator.ResonatorUpgrade> modes = ItemSkyResonator.getUpgrades(stack);
+			LuaList<String> modeNames = new LuaList<>(modes.size());
+
+			for (ItemSkyResonator.ResonatorUpgrade mode : modes) {
+				modeNames.add(mode.getUnlocalizedName());
+			}
+
+			return Collections.singletonMap("modes", modeNames.asMap()); */
+
+			return Collections.singletonMap("modes",
+				ItemSkyResonator.getUpgrades(context.getTarget()).stream()
+					.map(ItemSkyResonator.ResonatorUpgrade::getUnlocalizedName)
+					.collect(LuaList.toLuaList()).asMap());
+		}
+
+		@Nullable
+		@Override
+		public ItemStack getExample() {
+			return null;
+		}
+	};
+
+	public static final IMetaProvider<TileGrindstone> META_GRINDSTONE = new BaseMetaProvider<TileGrindstone>(
+		"Provides the item currently on this Grindstone"
+	) {
+
+		@Nonnull
+		@Override
+		public Map<String, Object> getMeta(@Nonnull IPartialContext<TileGrindstone> context) {
+			return Collections.singletonMap("item",
+				context.makePartialChild(context.getTarget().getGrindingItem()).getMeta());
+		}
+
+		@Nullable
+		@Override
+		public TileGrindstone getExample() {
+			//REFINE Implement this example
+			// Since `setGrindingItem` updates the BlockState, we will have to create the block for the TE too
+			return null;
+		}
+	};
+
+	public static final IMetaProvider<TileRitualPedestal> META_RITUAL_PEDESTAL= new BaseMetaProvider<TileRitualPedestal>() {
+
+		@Nonnull
+		@Override
+		public Map<String, Object> getMeta(@Nonnull IPartialContext<TileRitualPedestal> context) {
+			Map<String, Object> out = new HashMap<>(2);
+			TileRitualPedestal target = context.getTarget();
+
+			out.put("focus", context.makePartialChild(target.getInventoryHandler().getStackInSlot(0)).getMeta());
+
+			//Ritual constellation(s) provided by the crystal's meta
+
+			out.put("isWorking", target.isWorking());
+
+			return out;
+		}
+
+		@Nullable
+		@Override
+		public TileRitualPedestal getExample() {
+			return null;
+		}
+	};
+
+	public static final IMetaProvider<ItemStack> META_COLORED_LENS = new ItemStackContextMetaProvider<ItemColoredLens>(
+		ItemColoredLens.class,
+		"FIXME Set the description"
+	) {
+		@Nonnull
+		@Override
+		public Map<String, Object> getMeta(@Nonnull IPartialContext<ItemStack> context, @Nonnull ItemColoredLens item) {
+			ItemStack stack = context.getTarget();
+
+			// ... yay for the lack of a native reverse lookup or conversion based on the ordinal...
+			// (e.g. casting `int` to enum or a `TryParse` type method...)
```

Comment:
I miss .NET.

Reply:
`Enum.values[ordinal]`?

Reply:
That's what I'm resorting to, but I have to handle bounds checks myself.  And I dislike how some mod code just clamps the input to the enum bounds...

Reply:
`Optional<Enum> opt = Enum.values().stream().filter(e -> e.ordinal() == ordinal).findFirst()` /s